### PR TITLE
feat(service): add toVariant reverse extraction to EventVariantOf

### LINKED ISSUE #587: `EventVariantOf`: Add `toVariant` reverse extraction method
## Summary

Extend the `EventVariantOf` typeclass with a `toVariant` method that extracts a specific event variant from the event ADT. This is the reverse direction of the existing `fromVariant`.

Required by OutboundIntegration dispatch (#576) — the framework needs to go from the decoded event ADT back to a specific variant type to call typed handlers.

Part of #573 — Event Model DSL.

---

## Current State

Already merged via #575:

```haskell
class EventVariantOf event eventVariant where
  fromVariant :: eventVariant -> event

-- Identity instance:
instance EventVariantOf event event where
  fromVariant = identity
```

---

## Proposed Change

```haskell
class EventVariantOf event eventVariant where
  fromVariant :: eventVariant -> event
  toVariant   :: event -> Maybe eventVariant  -- NEW

-- Identity instance (extended):
instance {-# OVERLAPPABLE #-} EventVariantOf event event where
  fromVariant = identity
  toVariant   = Just
```

---

## Why This Is Needed

The OutboundIntegration dispatch (#576) decodes events from the EventStore as the full ADT (e.g., `CartEvent`), then needs to extract the specific variant (e.g., `ItemAdded`) to pass to a typed handler. `toVariant` provides this extraction:

```
JSON from EventStore
  → decode to CartEvent (full ADT)
  → toVariant @CartEvent @ItemAdded :: CartEvent → Maybe ItemAdded
  → Nothing = skip handler, Just variant = call handleEventImpl
```

---

## Example Instance

For standalone event types:

```haskell
data ItemAdded = ItemAdded
  { entityId :: Uuid, stockId :: Uuid, quantity :: Int }

instance EventVariantOf CartEvent ItemAdded where
  fromVariant e = CartItemAdded e.entityId e.stockId e.quantity
  toVariant event = case event of
    CartItemAdded eid sid qty ->
      Just (ItemAdded { entityId = eid, stockId = sid, quantity = qty })
    _ -> Nothing
```

---

## Scope

- Add `toVariant :: event -> Maybe eventVariant` to the `EventVariantOf` class
- Update the identity instance with `toVariant = Just`
- Update existing tests in `EventVariantOfSpec.hs` to cover `toVariant`
- Update existing `EventVariantOf` instances (if any beyond identity) to implement `toVariant`

---

## Dependencies

- Blocks **#576** — OutboundIntegration dispatch needs `toVariant`

---

## Affected Files

| Status | File | Purpose |
|--------|------|---------|
| Modified | `core/service/EventVariantOf.hs` | Add `toVariant` method + update identity instance |
| Modified | `core/test/EventVariantOfSpec.hs` | Add reverse extraction tests |

### PR DESCRIPTION
**Pull request by NickSeagull**
## Summary

Extends the `EventVariantOf` typeclass with a `toVariant :: event -> Maybe eventVariant` method — the reverse direction of the existing `fromVariant`. This enables extraction of specific event variants from the parent ADT, which is required by OutboundIntegration dispatch (#576) to route decoded events to typed handlers.

## Changes

- **`core/service/EventVariantOf.hs`**: Added `toVariant` method to the `EventVariantOf` class. Updated identity instance with `{-# OVERLAPPABLE #-}` pragma and `toVariant = Just`. Updated haddock documentation with `toVariant` examples.
- **`core/test/EventVariantOfSpec.hs`**: Added `toVariant` implementations to both custom test instances (`StandaloneCreated`, `StandaloneLabeled`). Added 15 new tests covering identity extraction, custom variant extraction (match + mismatch), boundary values (zero, negative, empty text), and `fromVariant`/`toVariant` round-trips.

## Checklist

- [x] Tests written (outside-in TDD)
- [x] `cabal build all` passes
- [x] `cabal test nhcore-test-core` passes (1016 examples, 0 failures)
- [x] `hlint` clean on changed files

Closes #587
Blocks #576

## Summary by CodeRabbit

* **New Features**
  * Added bidirectional event conversion capability, enabling events to be converted to and from typed variants with optional matching semantics.

* **Tests**
  * Added comprehensive test coverage for bidirectional event conversion including round-trip validation and pattern-matching scenarios.

### CODE REVIEW

File: core/test/EventVariantOfSpec.hs [hunk 1]
```diff
@@ -230,3 +240,101 @@ spec = do
...
+          let original = StandaloneLabeled { label = "" }
+          let adtEvent = fromVariant original :: TestEvent
+          let roundTripped = toVariant adtEvent :: Maybe StandaloneLabeled
+          roundTripped |> shouldBe (Just original)
+
+        -- Test 34: edge — round-trip identity
+        it "round-trip works for identity instance" \_ -> do
+          let original = TestEventA { testId = 42 }
+          let roundTripped = toVariant (fromVariant original :: TestEvent) :: Maybe TestEvent
+          roundTripped |> shouldBe (Just original)
```
[the remark this answers is not in the document]
Comment (NickSeagull) [reply core/test/EventVariantOfSpec.hs:340]: False positive. Haskell uses indentation-based scoping for `do`-blocks — there are no explicit closing braces or terminators. The file compiles cleanly (`cabal build all` passes) and all 1016 tests pass including the 15 new `toVariant` tests. The indentation structure is correct.
